xaitk_saliency/utils/tools/gen_detector_sal.py

```diff
@@ -0,0 +1,152 @@
+
+LOG = logging.getLogger(__name__)
+
+
+# mute logger to suppress verbose messages
+def _mute_log(_: Any) -> None: ...
+
+
+def gen_detector_sal(
+    ref_imgs: Collection[np.ndarray],
+    detector: DetectImageObjects,
+    img_perturber: PerturbImage,
+    sal_generator: GenerateDetectorProposalSaliency,
+    verbose: Optional[bool] = False,
+) -> Tuple[Collection[np.ndarray], Collection[np.ndarray]]:
+    """
+    Generate saliency maps for detections in each of the given images.
+    Detections generation, image perturbation, and saliency map creation is
+    performed by the implemenations of ``DetectImageObjects``,
+    ``PerturbImage``, and ``GenerateDetectorProposalSaliency`` that are passed,
+    respectively.
+
+    The resulting saliency maps matrix for each image has a shape of
+    [nDets, imgHeight, imgWidth].
+
+    :param ref_imgs: List of images to generate saliency maps for.
+    :param detector: Implementation of ``DetectImageObjects`` to generate
+        detections with.
+    :param img_perturber: Implementation of ``PerturbImage`` to use.
+    :param sal_generator: Implementation of
+        ``GenerateDetectorProposalSaliency`` to use.
+    :param verbose: Option to print out progress statements. Default is false.
+
+    :returns: Pair of parallel lists, the first being the computed saliency
+        maps, and the second being the reference detection bounding boxes for
+        each image. Each saliency map matrix has shape
+        [nDets, imgHeight, imgWidth] and each detections matrix [nDets, 4].
+    """
+
+    if verbose:
+        _log: Callable[[str], None] = LOG.info
+    else:
+        _log = _mute_log
+
+    _log("Getting reference detections.")
+    ref_dets_iter = detector(ref_imgs)
+
+    # List of saliency maps for all images
+    sal_maps_list = []
+    ref_dets_list = []
+
+    for i, (ref_img, ref_dets) in enumerate(zip(ref_imgs, ref_dets_iter)):
+
+        ref_dets = list(ref_dets)
+
+        num_dets = len(ref_dets)
+
+        _log(f'[{i+1}/{len(ref_imgs)}] {num_dets} reference detections found.')
+
+        # Return empty list if no reference detections are made
+        if num_dets == 0:
+            sal_maps_list.append(np.asarray([]).reshape(0, ref_img.shape[0], ref_img.shape[1]))
+            ref_dets_list.append(np.asarray([]).reshape(0, 4))
+            continue
+
+        _log(f'[{i+1}/{len(ref_imgs)}] Perturbing image.')
+
+        pert_masks = img_perturber(ref_img)
+        pert_imgs = occlude_image_batch(ref_img, pert_masks)
+
+        _log(f'[{i+1}/{len(ref_imgs)}] Getting perturbation detections.')
+
+        # Detections on all perturbed images
+        pert_dets_iter = detector(pert_imgs)
+
+        _log(f'[{i+1}/{len(ref_imgs)}] Generating saliency maps.')
+
+        # use keys of first reference detection as keys for rest of detections
+        class_dicts = [det[1] for det in ref_dets]
+        class_keys = list(class_dicts[0].keys())
```

Comment:
@schencej We were discussing last week how this could result in different key ordering from loop to loop, which would result in a bunch of detection matrices whose class confidence columns are not consistent/cross-applicable.
We should cache a `class_keys` across loops to mitigate this:
```
class_keys = None
for i, (ref_img, ref_dets) in enumerate(zip(ref_imgs, ref_dets_iter)):
    ...
    if class_keys is None:
        # class ordering based on first detection observed.
        # indexing into first detection -> classification dict (second position of tuple)
        class_keys = list(ref_dets[0][1].keys())
    # optionally error if any detection deviates from this set of keys
    # (something more performant than this)
    # if set(class_keys) != set(ref_dets[0][1].keys()):
    #     raise RuntimeError(...)
    ...
```

Reply:
You were correct @schencej and I was incorrect in my reading: the use of the perturbed detections are local to the loop and are not collected in the parent scope, so we do not need to align the keys across loops. I apologize for the confusion!